addon/mixins/contributor-mixin.js

```diff
@@ -0,0 +1,142 @@
+
+        var contributors = this.hasMany('contributors').hasManyRelationship;
+        this.set(
+            '_dirtyRelationships.contributors.update',
+            contributors.members.list.filter(m => !m.getRecord().get('isNew') && Object.keys(m.getRecord().changedAttributes()).length > 0)
+        );
+        // Need to included created contributors even in relationship
+        // hasLoaded is false
+        this.set(
+            '_dirtyRelationships.contributors.create',
+            contributors.members.list.filter(m => m.getRecord().get('isNew'))
+        );
+        // Contributors are a 'real' delete, not just a de-reference
+        this.set(
+            '_dirtyRelationships.contributors.delete',
+            this.get('_dirtyRelationships.contributors.remove') || []
+        );
+        this.set('_dirtyRelationships.contributors.remove', []);
+        return promise;
+    },
+    addContributor(userId, permission, isBibliographic, sendEmail, fullName, email) {
+        let contrib = this.store.createRecord('contributor', {
+            permission: permission,
+            bibliographic: isBibliographic,
+            sendEmail: sendEmail,
+            nodeId: this.get('id'),
+            userId: userId,
+            fullName: fullName,
+            email: email
+        });
+
+        return contrib.save();
+    },
+
+    addContributors(contributors, sendEmail) {
+        let payload = contributors.map(contrib => {
+            let c = this.store.createRecord('contributor', {
+                permission: contrib.permission,
+                bibliographic: contrib.bibliographic,
+                nodeId: this.get('id'),
+                userId: contrib.userId,
+                id: this.get('id') + '-' + contrib.userId,
+                unregisteredContributor: null
+            });
+            return c.serialize({
+                includeId: true,
+                includeUser: true
+            }).data;
+        });
+
+        let emailQuery = '';
+        if (!sendEmail) {
+            emailQuery = '?send_email=false';
+        } else if (sendEmail === 'preprint') {
+            emailQuery = '?send_email=preprint';
+        }
+
+        // TODO Get this working properly - should not be an ajax request in the future.
+        return this.store.adapterFor('contributor').ajax(this.get('links.relationships.contributors.links.related.href') + emailQuery, 'POST', {
+            data: {
+                data: payload
+            },
+            isBulk: true
+        }).then(resp => {
+            this.store.pushPayload(resp);
+            var createdContribs = Ember.A();
+            resp.data.map((contrib) => {
+                createdContribs.push(this.store.peekRecord('contributor', contrib.id));
+            });
+            return createdContribs;
+        });
+    },
+
+    removeContributor(contributor) {
+        return contributor.destroyRecord().then(rec => {
+            this.get('store')._removeFromIdMap(rec._internalModel);
```

Comment:
Could you clarify why a call to `_removeFromIdMap()` is needed here while it was not needed in the `removeContributor()` method of the `node` model previously?

Reply:
Ah, yes, I forgot to comment on this.  When attempting to remove the contributor from the preprint, and then add the contributor back, I was getting the following in the console: `Error: Assertion Failed: 'contributor' was saved to the server, but the response returned the new id '12345', which has already been used with another record.'` . It didn't seem like all evidence of the deleted contributor was being removed?  Seemed similar to this issue https://github.com/emberjs/data/issues/5014.

The only contributor-related difference I can think of between nodes and preprints are that nodes were using a `NodeMixin` to make these contributor requests to the ember-osf node model.  Preprints are calling the contributor methods on the ember-osf preprint model directly, but I don't know why this would make a difference.  I'll make a note to look into this more over the next couple days.  I don't want to call a private method, but haven't yet found a way around this.